{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, I have gone through my `targets.fa` file and cleaned the names of my sequences. All names contain only the letters `A-Z`, `a-z`, `0-9` and the underscore character `_`. For the sequences themselves, there are 2,700 or so from the Pfam (glycoside hydrolase family 1), about 150 from a paper that produced and purfied and tested the sequences in _E. coli_, and our lab's model BglB. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">P22505_BGLB_PAEPO\r\n",
      "MSENTFIFPATFMWGTSTSSYQIEGGTDEGGRTPSIWDTFCQIPGKVIGGDCGDVACDHF\r\n",
      "--\r\n",
      ">A0A0D3CKP2_BRAOL\r\n",
      "MRAFSHLFFSFLVILFAISCSDAFTRNDFPEEFLFGAATSAYQWEGAVDEDGRTPSVWDT\r\n",
      "--\r\n",
      ">A0A089PP77_PLUGE\r\n",
      "MNKLTLPKDFLWGGAVAAHQVEGGWNQGGKGPSICDVLTGGAHGVPREITQSVEAGKYYP\r\n"
     ]
    }
   ],
   "source": [
    "! grep '>' -A1 targets.fa | head -8"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I will check to see if there are any redundant sequences in the file, and remove them if so, by combining all the names together (don't lose data!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found sequence A0A022LM45_9MICO that is the same as another sequence\n",
      "Found sequence R6UYP4_9FIRM that is the same as another sequence\n",
      "Found sequence I1IAG4_BRADI that is the same as another sequence\n",
      "Found sequence M7ZDE0_TRIUA that is the same as another sequence\n",
      "Found sequence M7ZDE0_TRIUA that is the same as another sequence\n",
      "Found sequence A0A0D3GK12_9ORYZ that is the same as another sequence\n",
      "Found sequence A0A0D3GK12_9ORYZ that is the same as another sequence\n",
      "Found sequence W0AG68_9SPHN that is the same as another sequence\n",
      "Found sequence A9WDK4_CHLAA that is the same as another sequence\n",
      "Found sequence L0EFG7_THECK that is the same as another sequence\n",
      "Found sequence Q5YWH5_NOCFA that is the same as another sequence\n",
      "Found sequence D6ZED3_SEGRD that is the same as another sequence\n",
      "Found sequence H9JXC4_BOMMO that is the same as another sequence\n",
      "Found sequence R2PQY2_9ENTE that is the same as another sequence\n",
      "Found sequence K7L6U3_SOYBN that is the same as another sequence\n",
      "Found sequence K7L6U3_SOYBN that is the same as another sequence\n",
      "Found sequence A4FH46_SACEN that is the same as another sequence\n",
      "Found sequence A0A087GHZ2_ARAAL that is the same as another sequence\n",
      "Found sequence W5TKL6_9NOCA that is the same as another sequence\n",
      "Found sequence M4EMP4_BRARP that is the same as another sequence\n",
      "Found sequence G5AEN9_PHYSP that is the same as another sequence\n",
      "Found sequence R5QYD1_9FIRM that is the same as another sequence\n",
      "Found sequence E1ZE40_CHLVA that is the same as another sequence\n",
      "Found sequence M4ELT5_BRARP that is the same as another sequence\n",
      "Found sequence F7YX70_9THEM that is the same as another sequence\n",
      "Found sequence F6GUD1_VITVI that is the same as another sequence\n",
      "Found sequence D7KBD0_ARALL that is the same as another sequence\n",
      "Found sequence D7KBD0_ARALL that is the same as another sequence\n",
      "Found sequence B8BJH2_ORYSI that is the same as another sequence\n",
      "Found sequence W4Z6E7_STRPU that is the same as another sequence\n",
      "Found sequence W4Z6E7_STRPU that is the same as another sequence\n",
      "Found sequence K7M9A5_SOYBN that is the same as another sequence\n",
      "Found sequence H0EJW8_GLAL7 that is the same as another sequence\n",
      "Found sequence R6EFG1_9FIRM that is the same as another sequence\n",
      "Found sequence A8MCI6_CALMQ that is the same as another sequence\n",
      "Found sequence A8MCI6_CALMQ that is the same as another sequence\n",
      "Found sequence F6HB86_VITVI that is the same as another sequence\n",
      "Found sequence M4CKV2_BRARP that is the same as another sequence\n",
      "Found sequence H9JMQ8_BOMMO that is the same as another sequence\n",
      "Found sequence A0A0D3DFA9_BRAOL that is the same as another sequence\n",
      "Found sequence M0WGA6_HORVD that is the same as another sequence\n",
      "Found sequence M1AT72_SOLTU that is the same as another sequence\n",
      "Found sequence T1IGN2_RHOPR that is the same as another sequence\n",
      "Found sequence C5X5C0_SORBI that is the same as another sequence\n",
      "Found sequence W5TGN4_9NOCA that is the same as another sequence\n",
      "Found sequence T0N2G8_9EURY that is the same as another sequence\n",
      "Found sequence C5YTW7_SORBI that is the same as another sequence\n",
      "Found sequence C5YTW7_SORBI that is the same as another sequence\n",
      "Found sequence F6ICU5_9SPHN that is the same as another sequence\n",
      "Found sequence D4KCU5_9FIRM that is the same as another sequence\n",
      "Found sequence G3H8N4_CRIGR that is the same as another sequence\n",
      "Found sequence A0A087H863_ARAAL that is the same as another sequence\n",
      "Found sequence K1X514_MARBU that is the same as another sequence\n",
      "Found sequence W5WEG0_9PSEU that is the same as another sequence\n",
      "Found sequence W5TIN6_9NOCA that is the same as another sequence\n",
      "Found sequence G2T3V8_ROSHA that is the same as another sequence\n",
      "Found sequence R5RML7_9FIRM that is the same as another sequence\n",
      "Found sequence D8SVN7_SELML that is the same as another sequence\n",
      "Found sequence T0UQ96_9STRE that is the same as another sequence\n",
      "Found sequence R2QC65_9ENTE that is the same as another sequence\n",
      "Found sequence I1QK11_ORYGL that is the same as another sequence\n",
      "Found sequence M4C959_BRARP that is the same as another sequence\n",
      "Found sequence K3ZD93_SETIT that is the same as another sequence\n",
      "Found sequence R5R4K1_9FIRM that is the same as another sequence\n",
      "Found sequence A0A0D3HJ06_9ORYZ that is the same as another sequence\n",
      "Found sequence M7ZEP6_TRIUA that is the same as another sequence\n",
      "Found sequence D7ME83_ARALL that is the same as another sequence\n",
      "Found sequence G7IF68_MEDTR that is the same as another sequence\n",
      "Found sequence D0L6H9_GORB4 that is the same as another sequence\n",
      "Found sequence B9DS10_STRU0 that is the same as another sequence\n",
      "Found sequence B8HAF9_ARTCA that is the same as another sequence\n",
      "Found sequence R5R676_9FIRM that is the same as another sequence\n",
      "Found sequence R6KGG7_9CLOT that is the same as another sequence\n",
      "Found sequence A0A059MPX9_9NOCA that is the same as another sequence\n",
      "Found sequence M1MV39_9CLOT that is the same as another sequence\n",
      "Found sequence A0A0D3CG43_BRAOL that is the same as another sequence\n",
      "Found sequence R4T234_AMYOR that is the same as another sequence\n",
      "Found sequence F5XLS1_MICPN that is the same as another sequence\n",
      "Found sequence W5HP05_WHEAT that is the same as another sequence\n",
      "Found sequence E0STZ0_IGNAA that is the same as another sequence\n",
      "Found sequence Q74LJ7_LACJO that is the same as another sequence\n",
      "Found sequence S7Q2A9_GLOTA that is the same as another sequence\n",
      "Found sequence A9EWX4_SORC5 that is the same as another sequence\n",
      "Found sequence R5I063_9FIRM that is the same as another sequence\n",
      "Found sequence A8F6V8_PSELT that is the same as another sequence\n",
      "Found sequence A0A0D3B2T2_BRAOL that is the same as another sequence\n",
      "Found sequence D8T9S3_SELML that is the same as another sequence\n",
      "Found sequence H3F9R2_PRIPA that is the same as another sequence\n",
      "Found sequence H6R6W9_NOCCG that is the same as another sequence\n",
      "Found sequence W5THX4_9NOCA that is the same as another sequence\n",
      "Found sequence A8LFM3_FRASN that is the same as another sequence\n",
      "Found sequence A8LDF4_FRASN that is the same as another sequence\n",
      "Found sequence K0EUZ8_9NOCA that is the same as another sequence\n",
      "Found sequence D6X3T6_TRICA that is the same as another sequence\n",
      "Found sequence B0T913_CAUSK that is the same as another sequence\n",
      "Found sequence B0SWS2_CAUSK that is the same as another sequence\n",
      "Found sequence J9SBJ7_9ACTN that is the same as another sequence\n",
      "Found sequence R3TM72_9ENTE that is the same as another sequence\n",
      "Found sequence V4KH15_EUTSA that is the same as another sequence\n",
      "Found sequence F1TJ69_RHOHA that is the same as another sequence\n",
      "Found sequence I1IAG1_BRADI that is the same as another sequence\n",
      "Found sequence F4GH96_SPHCD that is the same as another sequence\n",
      "Found sequence Q5JDT7_THEKO that is the same as another sequence\n",
      "Found sequence BGAL_SULAC that is the same as another sequence\n",
      "Found sequence E2SEM7_9ACTN that is the same as another sequence\n",
      "Found sequence Q023T4_SOLUE that is the same as another sequence\n",
      "Found sequence BGL27_ORYSJ that is the same as another sequence\n",
      "Found sequence F6IEY3_9SPHN that is the same as another sequence\n",
      "Found sequence C5YAD8_SORBI that is the same as another sequence\n",
      "Found sequence E1QTH3_VULDI that is the same as another sequence\n",
      "Found sequence A0A0D3GK10_9ORYZ that is the same as another sequence\n",
      "Found sequence B8D213_HALOH that is the same as another sequence\n",
      "Found sequence M8ADE6_TRIUA that is the same as another sequence\n",
      "Found sequence M1CXV3_SOLTU that is the same as another sequence\n",
      "Found sequence G2P4X4_STRVO that is the same as another sequence\n",
      "Found sequence E7FHY4_PYRFU that is the same as another sequence\n",
      "Found sequence E7FHY4_PYRFU that is the same as another sequence\n",
      "Found sequence R5QPI3_9FIRM that is the same as another sequence\n",
      "Found sequence R4PXL6_9BACT that is the same as another sequence\n",
      "Found sequence R5TN47_9FIRM that is the same as another sequence\n",
      "Found sequence A5UXH8_ROSS1 that is the same as another sequence\n",
      "Found sequence R5QML1_9FIRM that is the same as another sequence\n",
      "Found sequence C3XVQ3_BRAFL that is the same as another sequence\n",
      "Found sequence A2Z317_ORYSI that is the same as another sequence\n",
      "Found sequence F6IDL9_9SPHN that is the same as another sequence\n",
      "Found sequence T0U4L2_9ENTE that is the same as another sequence\n",
      "Found sequence N4V3P5_FUSC1 that is the same as another sequence\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found sequence R5Q547_9FIRM that is the same as another sequence\n",
      "Found sequence A0A0D3EVD7_9ORYZ that is the same as another sequence\n",
      "Found sequence A0A0D3EVD7_9ORYZ that is the same as another sequence\n",
      "Found sequence E3IXX7_FRASU that is the same as another sequence\n",
      "Found sequence E3IUF1_FRASU that is the same as another sequence\n",
      "Found sequence I2F620_9BACT that is the same as another sequence\n",
      "Found sequence A0A0A1G1J2_9MYCO that is the same as another sequence\n",
      "Found sequence X2H7W6_9GAMM that is the same as another sequence\n",
      "Found sequence D1BJI3_SANKS that is the same as another sequence\n",
      "Found sequence C1A0M8_RHOE4 that is the same as another sequence\n",
      "Found sequence A0A087GIU9_ARAAL that is the same as another sequence\n",
      "Found sequence I1QYA3_ORYGL that is the same as another sequence\n",
      "Found sequence BGL28_ORYSJ that is the same as another sequence\n",
      "Found sequence K7LAN1_SOYBN that is the same as another sequence\n",
      "Found sequence Q9HIL6_THEAC that is the same as another sequence\n",
      "Found sequence C4LLI8_CORK4 that is the same as another sequence\n",
      "Found sequence H2YY90_CIOSA that is the same as another sequence\n",
      "Found sequence A7HNB8_FERNB that is the same as another sequence\n",
      "Found sequence B0RH42_CLAMS that is the same as another sequence\n",
      "Found sequence Q8U3U9_PYRFU that is the same as another sequence\n",
      "Found sequence B1W1N0_STRGG that is the same as another sequence\n",
      "Found sequence H9UG19_SPIAZ that is the same as another sequence\n",
      "Found sequence T0N4N1_9EURY that is the same as another sequence\n",
      "Found sequence C5Z8N4_SORBI that is the same as another sequence\n",
      "Found sequence Q5FKN4_LACAC that is the same as another sequence\n",
      "Found sequence D4BQN6_BIFBR that is the same as another sequence\n",
      "Found sequence D4KCX2_9FIRM that is the same as another sequence\n",
      "Found sequence R1B9G2_EMIHU that is the same as another sequence\n",
      "Found sequence M7U798_BOTF1 that is the same as another sequence\n",
      "Found sequence R7GJV3_9FIRM that is the same as another sequence\n",
      "Found sequence V6KHQ4_STRRC that is the same as another sequence\n",
      "Found sequence E0RXJ5_BUTPB that is the same as another sequence\n",
      "Found sequence A8F766_PSELT that is the same as another sequence\n",
      "Found sequence A0A0D3GK11_9ORYZ that is the same as another sequence\n",
      "Found sequence A0A0D3H202_9ORYZ that is the same as another sequence\n",
      "Found sequence H2J7V1_MARPK that is the same as another sequence\n",
      "Found sequence G2NFG8_STREK that is the same as another sequence\n",
      "Found sequence A8M9E1_CALMQ that is the same as another sequence\n",
      "Found sequence D6S2C4_9LACO that is the same as another sequence\n",
      "Found sequence H2J7W4_MARPK that is the same as another sequence\n",
      "Found sequence A0A087G3B0_ARAAL that is the same as another sequence\n",
      "Found sequence A0A087G3B0_ARAAL that is the same as another sequence\n",
      "Found sequence N2BYP6_9ACTN that is the same as another sequence\n",
      "Found sequence B1MDE8_MYCA9 that is the same as another sequence\n",
      "Found sequence Q0SHX5_RHOJR that is the same as another sequence\n",
      "Found sequence Q5FL27_LACAC that is the same as another sequence\n",
      "Found sequence S0KPG3_9ENTE that is the same as another sequence\n",
      "Found sequence A0A087GZB5_ARAAL that is the same as another sequence\n",
      "Found sequence A0A0D3H203_9ORYZ that is the same as another sequence\n",
      "Found sequence F5XGJ5_MICPN that is the same as another sequence\n",
      "Found sequence A0A088AS36_APIME that is the same as another sequence\n",
      "Found sequence Q2GAY4_NOVAD that is the same as another sequence\n",
      "Found sequence U5GQZ3_POPTR that is the same as another sequence\n",
      "Found sequence C8X761_NAKMY that is the same as another sequence\n",
      "Found sequence N4TYR9_FUSC1 that is the same as another sequence\n",
      "Found sequence F2NAZ1_CORGP that is the same as another sequence\n",
      "Found sequence M4F985_BRARP that is the same as another sequence\n",
      "Found sequence V5RWZ3_9BACT that is the same as another sequence\n",
      "Found sequence A0A096UGS3_MAIZE that is the same as another sequence\n",
      "Found sequence Q4JVR3_CORJK that is the same as another sequence\n",
      "Found sequence R5Q8V6_9FIRM that is the same as another sequence\n",
      "Found sequence I0XVK0_9LEPT that is the same as another sequence\n",
      "Found sequence B9GEM0_POPTR that is the same as another sequence\n",
      "Found sequence W5WQB6_9CORY that is the same as another sequence\n",
      "Found sequence I1NUA2_ORYGL that is the same as another sequence\n",
      "Found sequence K0EX11_9NOCA that is the same as another sequence\n",
      "Found sequence T0N308_9EURY that is the same as another sequence\n",
      "Found sequence K4AY09_SOLLC that is the same as another sequence\n",
      "Found sequence K4AY09_SOLLC that is the same as another sequence\n",
      "Found sequence G2YR88_BOTF4 that is the same as another sequence\n",
      "Found sequence A0A0D3EXD4_9ORYZ that is the same as another sequence\n",
      "Found sequence Q839A5_ENTFA that is the same as another sequence\n",
      "Found sequence M1ME05_9CLOT that is the same as another sequence\n",
      "Found sequence K0F927_9NOCA that is the same as another sequence\n",
      "Found sequence C9Z9I9_STRSW that is the same as another sequence\n",
      "Found sequence W5ESA4_WHEAT that is the same as another sequence\n",
      "Found sequence W5ESA4_WHEAT that is the same as another sequence\n",
      "Found sequence A0A0D1M1H3_9LACT that is the same as another sequence\n",
      "Found sequence D5U2M9_THEAM that is the same as another sequence\n",
      "Found sequence Q2GA89_NOVAD that is the same as another sequence\n",
      "Found sequence F9FB77_FUSOF that is the same as another sequence\n",
      "Found sequence I1R249_ORYGL that is the same as another sequence\n",
      "Found sequence A0A087GIU8_ARAAL that is the same as another sequence\n",
      "Found sequence A0A022QAL4_ERYGU that is the same as another sequence\n",
      "Found sequence A0A0D3CST2_BRAOL that is the same as another sequence\n",
      "Found sequence A0A0D3CST2_BRAOL that is the same as another sequence\n",
      "Found sequence D7LAV5_ARALL that is the same as another sequence\n",
      "Found sequence D7LAV5_ARALL that is the same as another sequence\n",
      "Found sequence BGAL_SULSO that is the same as another sequence\n",
      "Found sequence D9PZ08_ACIS3 that is the same as another sequence\n",
      "Found sequence D0NXN3_PHYIT that is the same as another sequence\n",
      "Found sequence L5LF94_MYODS that is the same as another sequence\n",
      "Found sequence F7GB62_MACMU that is the same as another sequence\n",
      "Found sequence H6MS22_GORPV that is the same as another sequence\n",
      "Found sequence A0A0D3BSN1_BRAOL that is the same as another sequence\n",
      "Found sequence A0A0D3EY85_9ORYZ that is the same as another sequence\n",
      "Found sequence D4IYQ1_BUTFI that is the same as another sequence\n",
      "Found sequence A0A087GIV0_ARAAL that is the same as another sequence\n",
      "Found sequence D8QN60_SELML that is the same as another sequence\n",
      "Found sequence A0A087HEU1_ARAAL that is the same as another sequence\n",
      "Found sequence V4SJI5_STRIN that is the same as another sequence\n",
      "Found sequence A0A073CLU2_PLAAG that is the same as another sequence\n",
      "Found sequence D8R716_SELML that is the same as another sequence\n",
      "Found sequence A0A091DIZ4_FUKDA that is the same as another sequence\n",
      "Found sequence A0A091DIZ4_FUKDA that is the same as another sequence\n",
      "Found sequence M7ZZX2_TRIUA that is the same as another sequence\n",
      "Found sequence D8LQW6_ECTSI that is the same as another sequence\n",
      "Found sequence D8LQW6_ECTSI that is the same as another sequence\n",
      "Found sequence Q6KZ14_PICTO that is the same as another sequence\n",
      "Found sequence A0A096SS60_MAIZE that is the same as another sequence\n",
      "Found sequence M4CKV4_BRARP that is the same as another sequence\n",
      "Found sequence E1ZYR2_CAMFO that is the same as another sequence\n",
      "Found sequence G3Y786_ASPNA that is the same as another sequence\n",
      "Found sequence A0A0D3DLX2_BRAOL that is the same as another sequence\n",
      "Found sequence R5QE22_9FIRM that is the same as another sequence\n",
      "Found sequence A2Y3U4_ORYSI that is the same as another sequence\n",
      "Found sequence A2Y3U4_ORYSI that is the same as another sequence\n",
      "Found sequence R5RF44_9FIRM that is the same as another sequence\n",
      "Found sequence D7LL70_ARALL that is the same as another sequence\n",
      "Found sequence SDJGI_31 that is the same as another sequence\n",
      "Found sequence SDJGI_42 that is the same as another sequence\n",
      "Found sequence SDJGI_34 that is the same as another sequence\n",
      "Found sequence SDJGI_33 that is the same as another sequence\n",
      "Found sequence SDJGI_83 that is the same as another sequence\n",
      "Found sequence SDJGI_136 that is the same as another sequence\n",
      "Found sequence SDJGI_86 that is the same as another sequence\n",
      "Found sequence SDJGI_141 that is the same as another sequence\n",
      "Found sequence SDJGI_181 that is the same as another sequence\n",
      "Found sequence SDJGI_138 that is the same as another sequence\n",
      "Found sequence SDJGI_82 that is the same as another sequence\n",
      "Found sequence SDJGI_78 that is the same as another sequence\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found sequence SDJGI_87 that is the same as another sequence\n",
      "Found sequence SDJGI_183 that is the same as another sequence\n",
      "Found sequence SDJGI_30 that is the same as another sequence\n",
      "Found sequence SDJGI_40 that is the same as another sequence\n",
      "Found sequence SDJGI_53 that is the same as another sequence\n",
      "Found sequence SDJGI_170 that is the same as another sequence\n",
      "Found sequence SDJGI_52 that is the same as another sequence\n",
      "Found sequence SDJGI_197 that is the same as another sequence\n",
      "Found sequence SDJGI_73 that is the same as another sequence\n",
      "Found sequence SDJGI_128 that is the same as another sequence\n",
      "Found sequence SDJGI_151 that is the same as another sequence\n",
      "Found sequence SDJGI_177 that is the same as another sequence\n",
      "Found sequence SDJGI_117 that is the same as another sequence\n",
      "Found sequence SDJGI_55 that is the same as another sequence\n",
      "Found sequence SDJGI_112 that is the same as another sequence\n",
      "Found sequence SDJGI_163 that is the same as another sequence\n",
      "Found sequence SDJGI_67 that is the same as another sequence\n",
      "Found sequence SDJGI_90 that is the same as another sequence\n",
      "Found sequence SDJGI_44 that is the same as another sequence\n",
      "Found sequence SDJGI_101 that is the same as another sequence\n",
      "Found sequence SDJGI_95 that is the same as another sequence\n",
      "Found sequence SDJGI_121 that is the same as another sequence\n",
      "Found sequence SDJGI_178 that is the same as another sequence\n",
      "Found sequence SDJGI_100 that is the same as another sequence\n",
      "Found sequence SDJGI_150 that is the same as another sequence\n",
      "Found sequence SDJGI_60 that is the same as another sequence\n",
      "Found sequence SDJGI_109 that is the same as another sequence\n",
      "Found sequence SDJGI_179 that is the same as another sequence\n",
      "Found sequence SDJGI_77 that is the same as another sequence\n",
      "Found sequence SDJGI_43 that is the same as another sequence\n",
      "Found sequence SDJGI_92 that is the same as another sequence\n",
      "Found sequence SDJGI_131 that is the same as another sequence\n",
      "Found sequence SDJGI_74 that is the same as another sequence\n",
      "Found sequence SDJGI_126 that is the same as another sequence\n",
      "Found sequence SDJGI_75 that is the same as another sequence\n",
      "Found sequence SDJGI_64 that is the same as another sequence\n",
      "Found sequence SDJGI_152 that is the same as another sequence\n",
      "Found sequence SDJGI_125 that is the same as another sequence\n",
      "Found sequence SDJGI_132 that is the same as another sequence\n",
      "Found sequence SDJGI_153 that is the same as another sequence\n",
      "Found sequence SDJGI_157 that is the same as another sequence\n",
      "Found sequence SDJGI_22 that is the same as another sequence\n",
      "Found sequence SDJGI_47 that is the same as another sequence\n",
      "Found sequence SDJGI_54 that is the same as another sequence\n",
      "Found sequence SDJGI_63 that is the same as another sequence\n",
      "Found sequence SDJGI_76 that is the same as another sequence\n",
      "Found sequence SDJGI_45 that is the same as another sequence\n",
      "Found sequence SDJGI_36 that is the same as another sequence\n",
      "Found sequence SDJGI_175 that is the same as another sequence\n",
      "Found sequence SDJGI_16 that is the same as another sequence\n",
      "Found sequence SDJGI_17 that is the same as another sequence\n",
      "Found sequence SDJGI_103 that is the same as another sequence\n",
      "Found sequence SDJGI_110 that is the same as another sequence\n",
      "Found sequence SDJGI_113 that is the same as another sequence\n",
      "Found sequence SDJGI_115 that is the same as another sequence\n",
      "Found sequence SDJGI_122 that is the same as another sequence\n",
      "Found sequence SDJGI_130 that is the same as another sequence\n",
      "Found sequence SDJGI_144 that is the same as another sequence\n",
      "Found sequence SDJGI_145 that is the same as another sequence\n",
      "Found sequence SDJGI_146 that is the same as another sequence\n",
      "Found sequence SDJGI_147 that is the same as another sequence\n",
      "Found sequence SDJGI_191 that is the same as another sequence\n",
      "Found sequence SDJGI_88 that is the same as another sequence\n",
      "Found sequence SDJGI_89 that is the same as another sequence\n",
      "Found sequence SDJGI_99 that is the same as another sequence\n",
      "Found sequence SDJGI_142 that is the same as another sequence\n",
      "Found sequence SDJGI_143 that is the same as another sequence\n",
      "Found sequence SDJGI_39 that is the same as another sequence\n",
      "Found sequence SDJGI_41 that is the same as another sequence\n",
      "Found sequence SDJGI_38 that is the same as another sequence\n",
      "Found sequence SDJGI_114 that is the same as another sequence\n",
      "Found sequence SDJGI_133 that is the same as another sequence\n",
      "Found sequence SDJGI_119 that is the same as another sequence\n",
      "Found sequence SDJGI_186 that is the same as another sequence\n",
      "Found sequence SDJGI_173 that is the same as another sequence\n",
      "Found sequence SDJGI_194 that is the same as another sequence\n",
      "Found sequence SDJGI_94 that is the same as another sequence\n",
      "Found sequence SDJGI_97 that is the same as another sequence\n",
      "Found sequence SDJGI_48 that is the same as another sequence\n",
      "Found sequence SDJGI_59 that is the same as another sequence\n",
      "Found sequence SDJGI_23 that is the same as another sequence\n",
      "Found sequence SDJGI_118 that is the same as another sequence\n",
      "Found sequence SDJGI_124 that is the same as another sequence\n",
      "Found sequence SDJGI_155 that is the same as another sequence\n",
      "Found sequence SDJGI_164 that is the same as another sequence\n",
      "Found sequence SDJGI_172 that is the same as another sequence\n",
      "Found sequence SDJGI_174 that is the same as another sequence\n",
      "Found sequence SDJGI_46 that is the same as another sequence\n",
      "Found sequence SDJGI_184 that is the same as another sequence\n",
      "Found sequence SDJGI_72 that is the same as another sequence\n",
      "Found sequence SDJGI_116 that is the same as another sequence\n",
      "Found sequence Q59976 that is the same as another sequence\n",
      "Found sequence Q7MG41 that is the same as another sequence\n",
      "Found sequence Q97AX4 that is the same as another sequence\n",
      "Found sequence Q59976 that is the same as another sequence\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO \n",
    "\n",
    "seen = []\n",
    "for record in SeqIO.parse('targets.fa', 'fasta'):\n",
    "    if record.seq in seen:\n",
    "        print('Found sequence {} that is the same as another sequence'.format(record.id))\n",
    "        continue \n",
    "    seen.append(record.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Well, that is interesting. There is a lot of redundancy. Instead of blindly discarding, I'd like to keep the name from the JGI data set, because we have experimental data for those sequences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dog mouse\n"
     ]
    }
   ],
   "source": [
    "def longer(x, y):\n",
    "    '''\n",
    "    returns the longer of iterables x, y\n",
    "    if they are the same length, returns y \n",
    "    '''\n",
    "    if x > y:\n",
    "        return x\n",
    "    return y\n",
    "\n",
    "print(longer('cat', 'dog'), longer('cat', 'mouse'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2533"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO \n",
    "\n",
    "seen = []\n",
    "wanted = []\n",
    "for record in SeqIO.parse('targets.fa', 'fasta'):\n",
    "    # do I want you to appear in the final alignment? \n",
    "    if record.seq not in seen:\n",
    "        seen.append(record.seq)\n",
    "        wanted.append(record)\n",
    "    else:\n",
    "        # I have seen this seq\n",
    "        # find the better name \n",
    "        for w in wanted:\n",
    "            if w.seq == record.seq:\n",
    "                # they are same sequence\n",
    "                w.id = longer(w.id, record.id) \n",
    "                \n",
    "len(wanted) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2533"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SeqIO.write(wanted, 'target.fa', 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OK, so now we have 2,533 non-redundant sequences in our FASTA file. I will now align them using Mafft via the following commmand \n",
    "\n",
    "```bash\n",
    "mafft --auto --thread -1 target.fa > mafft_out.fa\n",
    "```\n",
    "\n",
    "(that takes <1 min on my dual core 1.3 GHz MacBook Air), and also with Muscle \n",
    "\n",
    "```bash\n",
    "muscle -in target.fa > muscle_out.fa \n",
    "```\n",
    "\n",
    "Tree building\n",
    "\n",
    "For tree building, I used FastTree with the following command line \n",
    "\n",
    "```bash \n",
    "FastTree mafft_out.fa > mafft_tree.nwk \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we have the unpleasant task of trying to use Graphlan. Why it is so slow and can't just be `import`ed is a mystery. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bglb SDJGI_16\n",
      "run_A SDJGI_142\n",
      "run_B SDJGI_139\n",
      "run_C SDJGI_4\n"
     ]
    }
   ],
   "source": [
    "for w in wanted: \n",
    "    for record in SeqIO.parse('data.fa', 'fasta'):\n",
    "        if record.seq == w.seq:\n",
    "            print(record.id, w.id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's draw a tree "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sorting</th>\n",
       "      <th>Organism</th>\n",
       "      <th>PDB</th>\n",
       "      <th>Kindgom</th>\n",
       "      <th>Sol_exp_(mg/ml)</th>\n",
       "      <th>Cellobiose_binary(Act)</th>\n",
       "      <th>Cellobiose_therm_binary</th>\n",
       "      <th>pH7_40</th>\n",
       "      <th>pH4_40</th>\n",
       "      <th>pH7_60</th>\n",
       "      <th>pH7_80</th>\n",
       "      <th>pH4_80</th>\n",
       "      <th>pH7_80_pre80</th>\n",
       "      <th>pH7_90_pre90</th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>organism</th>\n",
       "      <th>uniprot</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SDJGI_31</th>\n",
       "      <td>1</td>\n",
       "      <td>Picrophilus torridus DSM 9790</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>0.010014</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.043767</td>\n",
       "      <td>0.042467</td>\n",
       "      <td>0.042100</td>\n",
       "      <td>0.040767</td>\n",
       "      <td>0.041400</td>\n",
       "      <td>0.04085</td>\n",
       "      <td>0.041333</td>\n",
       "      <td>SDJGI_31</td>\n",
       "      <td>MLPKNFLLGFSLAGFQSEMGISDPDSNSDWWLWVHDPVNIRTGLVS...</td>\n",
       "      <td>Picrophilus torridus</td>\n",
       "      <td>Q6KZ14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SDJGI_139</th>\n",
       "      <td>2</td>\n",
       "      <td>Thermoplasma volcanium GSS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>0.399383</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.111767</td>\n",
       "      <td>1.494767</td>\n",
       "      <td>1.898033</td>\n",
       "      <td>1.949733</td>\n",
       "      <td>1.498800</td>\n",
       "      <td>2.11665</td>\n",
       "      <td>0.045933</td>\n",
       "      <td>SDJGI_139</td>\n",
       "      <td>MVENNFPEDFKFGWSQSGFQSEMGYDNAMDDKSDWYVWVHDKENIQ...</td>\n",
       "      <td>Thermoplasma volcanium</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SDJGI_35</th>\n",
       "      <td>3</td>\n",
       "      <td>Magnaporthe grisea Y34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryote</td>\n",
       "      <td>0.012451</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.040667</td>\n",
       "      <td>0.040733</td>\n",
       "      <td>0.040500</td>\n",
       "      <td>0.040533</td>\n",
       "      <td>0.042033</td>\n",
       "      <td>0.04080</td>\n",
       "      <td>0.041333</td>\n",
       "      <td>SDJGI_35</td>\n",
       "      <td>MSLPKDFLWGFATASYQIEGAIDKDGRGPSIWDTFTAIPGKVADGS...</td>\n",
       "      <td>Magnaporthe grisea</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SDJGI_137</th>\n",
       "      <td>4</td>\n",
       "      <td>Thermococcus kodakarensis KOD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>0.004728</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.041167</td>\n",
       "      <td>0.041233</td>\n",
       "      <td>0.040867</td>\n",
       "      <td>0.041000</td>\n",
       "      <td>0.042533</td>\n",
       "      <td>0.04190</td>\n",
       "      <td>0.042100</td>\n",
       "      <td>SDJGI_137</td>\n",
       "      <td>MREFAWGVVQSAFQFEMGDPLRRFIDTRTDWWHWVRDPLNIKNDLV...</td>\n",
       "      <td>Thermococcus kodakarensis</td>\n",
       "      <td>Q5JDV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SDJGI_42</th>\n",
       "      <td>5</td>\n",
       "      <td>Saccharophagus degradans 2-40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>0.281507</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.795000</td>\n",
       "      <td>0.701667</td>\n",
       "      <td>1.006167</td>\n",
       "      <td>1.276900</td>\n",
       "      <td>0.511700</td>\n",
       "      <td>0.04050</td>\n",
       "      <td>0.041500</td>\n",
       "      <td>SDJGI_42</td>\n",
       "      <td>MNRLTLPPSSRLRSKEFTFGVATSSYQIEGGIDSRLPCNWDTFCEQ...</td>\n",
       "      <td>Saccharophagus degradans</td>\n",
       "      <td>Q21KX3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sorting                        Organism  PDB    Kindgom  \\\n",
       "gene                                                                 \n",
       "SDJGI_31         1   Picrophilus torridus DSM 9790  NaN    Archaea   \n",
       "SDJGI_139        2     Thermoplasma volcanium GSS1  NaN    Archaea   \n",
       "SDJGI_35         3          Magnaporthe grisea Y34  NaN  Eukaryote   \n",
       "SDJGI_137        4  Thermococcus kodakarensis KOD1  NaN    Archaea   \n",
       "SDJGI_42         5   Saccharophagus degradans 2-40  NaN   Bacteria   \n",
       "\n",
       "           Sol_exp_(mg/ml)  Cellobiose_binary(Act)  Cellobiose_therm_binary  \\\n",
       "gene                                                                          \n",
       "SDJGI_31          0.010014                     0.0                      0.0   \n",
       "SDJGI_139         0.399383                     1.0                      1.0   \n",
       "SDJGI_35          0.012451                     0.0                      0.0   \n",
       "SDJGI_137         0.004728                     0.0                      0.0   \n",
       "SDJGI_42          0.281507                     1.0                      0.0   \n",
       "\n",
       "             pH7_40    pH4_40    pH7_60    pH7_80    pH4_80  pH7_80_pre80  \\\n",
       "gene                                                                        \n",
       "SDJGI_31   0.043767  0.042467  0.042100  0.040767  0.041400       0.04085   \n",
       "SDJGI_139  1.111767  1.494767  1.898033  1.949733  1.498800       2.11665   \n",
       "SDJGI_35   0.040667  0.040733  0.040500  0.040533  0.042033       0.04080   \n",
       "SDJGI_137  0.041167  0.041233  0.040867  0.041000  0.042533       0.04190   \n",
       "SDJGI_42   1.795000  0.701667  1.006167  1.276900  0.511700       0.04050   \n",
       "\n",
       "           pH7_90_pre90       Name  \\\n",
       "gene                                 \n",
       "SDJGI_31       0.041333   SDJGI_31   \n",
       "SDJGI_139      0.045933  SDJGI_139   \n",
       "SDJGI_35       0.041333   SDJGI_35   \n",
       "SDJGI_137      0.042100  SDJGI_137   \n",
       "SDJGI_42       0.041500   SDJGI_42   \n",
       "\n",
       "                                                    Sequence  \\\n",
       "gene                                                           \n",
       "SDJGI_31   MLPKNFLLGFSLAGFQSEMGISDPDSNSDWWLWVHDPVNIRTGLVS...   \n",
       "SDJGI_139  MVENNFPEDFKFGWSQSGFQSEMGYDNAMDDKSDWYVWVHDKENIQ...   \n",
       "SDJGI_35   MSLPKDFLWGFATASYQIEGAIDKDGRGPSIWDTFTAIPGKVADGS...   \n",
       "SDJGI_137  MREFAWGVVQSAFQFEMGDPLRRFIDTRTDWWHWVRDPLNIKNDLV...   \n",
       "SDJGI_42   MNRLTLPPSSRLRSKEFTFGVATSSYQIEGGIDSRLPCNWDTFCEQ...   \n",
       "\n",
       "                            organism uniprot  \n",
       "gene                                          \n",
       "SDJGI_31        Picrophilus torridus  Q6KZ14  \n",
       "SDJGI_139     Thermoplasma volcanium     NaN  \n",
       "SDJGI_35          Magnaporthe grisea     NaN  \n",
       "SDJGI_137  Thermococcus kodakarensis  Q5JDV2  \n",
       "SDJGI_42    Saccharophagus degradans  Q21KX3  "
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x11baf10f0>"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAD4VJREFUeJzt3X+s3Xddx/Hny5XJj6vrZvGm6cCiqeB0ouyKEwi5tRhg\nI24mZA4nFLKkMSAsBiOFP9wfhqQkzoBBNA0QaiTczLG4KoguxQMuuGELg/2ouEm70dGtAtvw1ATT\n7e0f95TcbO3u6fmee0/v5zwfSXPP9+fn/b63ed3P/dxzzk1VIUlq149MugBJ0soy6CWpcQa9JDXO\noJekxhn0ktQ4g16SGmfQS1LjDHpJapxBL0mNWzfpAgA2bNhQmzdvHuna48eP87znPW+8Ba0R09q7\nfU+Xae0blu/9wIED36mq5y93n7Mi6Ddv3sz+/ftHurbX6zE/Pz/egtaIae3dvqfLtPYNy/ee5IFh\n7uPSjSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNe6seGVsV5t3fmYi4x7e\ndflExpWkM+GMXpIaZ9BLUuMMeklq3LJBn+TjSY4luXvJvguS3JrkvsHH85cce2+S+5N8I8lrV6pw\nSdJwhpnRfwJ43VP27QT2VdUWYN9gmyQXAVcDPz+45iNJzhlbtZKkM7Zs0FfVF4HvPWX3FcCeweM9\nwJVL9i9U1Q+q6hBwP/DyMdUqSRrBqGv0s1V1dPD4YWB28HgT8K0l5x0Z7JMkTUjn59FXVSWpM70u\nyQ5gB8Ds7Cy9Xm+k8fv9Pu+++ImRru1q1JrHpd/vT7yGSbDv6TKtfcP4eh816B9JsrGqjibZCBwb\n7H8IeMGS8y4c7HuaqtoN7AaYm5urUf9UWK/X44bbjo90bVeHr5mfyLgnTeufWLPv6TKtfcP4eh91\n6WYvsH3weDtwy5L9Vyf50SQvArYAX+5WoiSpi2Vn9Ek+BcwDG5IcAa4HdgE3JrkWeAC4CqCq7kly\nI3AvcAJ4R1VNZl1FkgQMEfRV9abTHNp2mvPfD7y/S1GSpPHxlbGS1DiDXpIaZ9BLUuMMeklqnEEv\nSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLU\nOINekhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNc6gl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z\n6CWpcQa9JDWuU9An+YMk9yS5O8mnkjw7yQVJbk1y3+Dj+eMqVpJ05kYO+iSbgHcBc1X1C8A5wNXA\nTmBfVW0B9g22JUkT0nXpZh3wnCTrgOcC3wauAPYMju8Bruw4hiSpg5GDvqoeAv4UeBA4CjxeVf8M\nzFbV0cFpDwOznauUJI0sVTXahYtr758Gfht4DPhb4Cbgw1W1fsl5j1bV09bpk+wAdgDMzs5esrCw\nMFId/X6fQ48/MdK1XV286byJjHtSv99nZmZmojVMgn1Pl2ntG5bvfevWrQeqam65+6zrUMNrgENV\n9d8ASW4GXgE8kmRjVR1NshE4dqqLq2o3sBtgbm6u5ufnRyqi1+txw23HR7q2q8PXzE9k3JN6vR6j\nft7WMvueLtPaN4yv9y5r9A8ClyZ5bpIA24CDwF5g++Cc7cAt3UqUJHUx8oy+qu5IchPwFeAE8FUW\nZ+gzwI1JrgUeAK4aR6GSpNF0Wbqhqq4Hrn/K7h+wOLuXJJ0FfGWsJDXOoJekxhn0ktQ4g16SGmfQ\nS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjTPoJalxBr0k\nNc6gl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1Lj\nDHpJapxBL0mN6xT0SdYnuSnJfyQ5mOTXklyQ5NYk9w0+nj+uYiVJZ67rjP5DwOeq6iXAS4GDwE5g\nX1VtAfYNtiVJEzJy0Cc5D3g18DGAqvq/qnoMuALYMzhtD3Bl1yIlSaNLVY12YfJLwG7gXhZn8weA\n64CHqmr94JwAj57cfsr1O4AdALOzs5csLCyMVEe/3+fQ40+MdG1XF286byLjntTv95mZmZloDZNg\n39NlWvuG5XvfunXrgaqaW+4+XYJ+DrgdeGVV3ZHkQ8D3gXcuDfYkj1bVM67Tz83N1f79+0eqo9fr\n8dbPHR/p2q4O77p8IuOe1Ov1mJ+fn2gNk2Df02Va+4ble08yVNB3WaM/AhypqjsG2zcBLwMeSbJx\nUMRG4FiHMSRJHY0c9FX1MPCtJC8e7NrG4jLOXmD7YN924JZOFUqSOlnX8fp3Ap9Mci7wTeBtLH7z\nuDHJtcADwFUdx5AkddAp6KvqTuBU60PbutxXkjQ+vjJWkhpn0EtS4wx6SWqcQS9JjTPoJalxBr0k\nNc6gl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1Lj\nDHpJapxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNa5z\n0Cc5J8lXk/zDYPuCJLcmuW/w8fzuZUqSRjWOGf11wMEl2zuBfVW1Bdg32JYkTUinoE9yIXA58NEl\nu68A9gwe7wGu7DKGJKmbrjP6DwJ/BDy5ZN9sVR0dPH4YmO04hiSpg1TVaBcmbwAuq6q3J5kH/rCq\n3pDksapav+S8R6vqaev0SXYAOwBmZ2cvWVhYGKmOfr/PocefGOnari7edN5Exj2p3+8zMzMz0Rom\nwb6ny7T2Dcv3vnXr1gNVNbfcfdZ1qOGVwG8muQx4NvDjSf4GeCTJxqo6mmQjcOxUF1fVbmA3wNzc\nXM3Pz49URK/X44bbjo90bVeHr5mfyLgn9Xo9Rv28rWX2PV2mtW8YX+8jL91U1Xur6sKq2gxcDXy+\nqn4X2AtsH5y2Hbilc5WSpJGtxPPodwG/keQ+4DWDbUnShHRZuvmhquoBvcHj7wLbxnFfSVJ3vjJW\nkhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNc6gl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWp\ncQa9JDXOoJekxhn0ktQ4g16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLUOINekhpn\n0EtS4wx6SWqcQS9JjTPoJalxBr0kNc6gl6TGjRz0SV6Q5F+S3JvkniTXDfZfkOTWJPcNPp4/vnIl\nSWeqy4z+BPDuqroIuBR4R5KLgJ3AvqraAuwbbEuSJmTkoK+qo1X1lcHj/wEOApuAK4A9g9P2AFd2\nLVKSNLqxrNEn2Qz8MnAHMFtVRweHHgZmxzGGJGk0qapuN0hmgC8A76+qm5M8VlXrlxx/tKqetk6f\nZAewA2B2dvaShYWFkcbv9/scevyJ0Yrv6OJN501k3JP6/T4zMzMTrWES7Hu6TGvfsHzvW7duPVBV\nc8vdZ12XIpI8C/g08Mmqunmw+5EkG6vqaJKNwLFTXVtVu4HdAHNzczU/Pz9SDb1ejxtuOz7StV0d\nvmZ+IuOe1Ov1GPXztpbZ93SZ1r5hfL13edZNgI8BB6vqz5Yc2gtsHzzeDtwyenmSpK66zOhfCbwZ\nuCvJnYN97wN2ATcmuRZ4ALiqW4mSpC5GDvqqug3IaQ5vG/W+kqTx8pWxktQ4g16SGmfQS1LjDHpJ\napxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLUuE5vajbtNu/8zETGPbzr8omMK2ltckYvSY0z6CWpcQa9\nJDXONfo16OTvBt598QneOqHfE6w2fy8hjc4ZvSQ1zhm91oSlz3Ba7Z9k/GlCa50zeklqnDN66Sz1\n1NdprOZPMv4U0xZn9JLUOINekhpn0EtS4wx6SWqcQS9JjfNZN9IyJvUupdK4OKOXpMYZ9JLUOINe\nkhpn0EtS4wx6SWqcz7qR9DT+PeS2OKOXpMat2Iw+yeuADwHnAB+tql0rNZYkdTHJ10qsxk8xKzKj\nT3IO8BfA64GLgDcluWglxpIkPbOVmtG/HLi/qr4JkGQBuAK4d4XGk9SAU82sp+lvI6+UlVqj3wR8\na8n2kcE+SdIqm9izbpLsAHYMNvtJvjHirTYA3xlPVWvLu6a0d/ueLq33nQ884+Hlev+pYcZYqaB/\nCHjBku0LB/t+qKp2A7u7DpRkf1XNdb3PWjStvdv3dJnWvmF8va/U0s2/A1uSvCjJucDVwN4VGkuS\n9AxWZEZfVSeS/D7wTyw+vfLjVXXPSowlSXpmK7ZGX1WfBT67UvdfovPyzxo2rb3b93SZ1r5hTL2n\nqsZxH0nSWcq3QJCkxq2ZoE/yuiTfSHJ/kp2nOJ4kfz44/vUkL5tEneM2RN/XDPq9K8mXkrx0EnWu\nhOV6X3LeryQ5keSNq1nfShmm7yTzSe5Mck+SL6x2jSthiP/r5yX5+yRfG/T9tknUOW5JPp7kWJK7\nT3O8e7ZV1Vn/j8Vf6P4X8NPAucDXgIuecs5lwD8CAS4F7ph03avU9yuA8wePX99C38P2vuS8z7P4\n+6A3TrruVfqar2fxVeYvHGz/5KTrXqW+3wd8YPD4+cD3gHMnXfsYen818DLg7tMc75xta2VG/8O3\nVKiq/wNOvqXCUlcAf12LbgfWJ9m42oWO2bJ9V9WXqurRwebtLL5moQXDfM0B3gl8Gji2msWtoGH6\n/h3g5qp6EKCqWuh9mL4L+LEkAWZYDPoTq1vm+FXVF1ns5XQ6Z9taCfph3lKhxbddONOermXxO38L\nlu09ySbgt4C/XMW6VtowX/OfBc5P0ktyIMlbVq26lTNM3x8Gfg74NnAXcF1VPbk65U1U52zzD480\nIslWFoP+VZOuZRV9EHhPVT25OMmbGuuAS4BtwHOAf0tye1X952TLWnGvBe4Efh34GeDWJP9aVd+f\nbFlnv7US9Mu+pcKQ56w1Q/WU5BeBjwKvr6rvrlJtK22Y3ueAhUHIbwAuS3Kiqv5udUpcEcP0fQT4\nblUdB44n+SLwUmAtB/0wfb8N2FWLC9f3JzkEvAT48uqUODGds22tLN0M85YKe4G3DH5DfSnweFUd\nXe1Cx2zZvpO8ELgZeHNjM7ple6+qF1XV5qraDNwEvH2NhzwM93/9FuBVSdYleS7wq8DBVa5z3Ibp\n+0EWf4ohySzwYuCbq1rlZHTOtjUxo6/TvKVCkt8bHP8rFp91cRlwP/C/LH73X9OG7PuPgZ8APjKY\n2Z6oBt4AasjemzNM31V1MMnngK8DT7L4F9xO+dS8tWLIr/efAJ9IcheLz0B5T1Wt+Xe1TPIpYB7Y\nkOQIcD3wLBhftvnKWElq3FpZupEkjcigl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWpcf8P\nkeiOf8okO9EAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11bb12be0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fr['Sol_exp_(mg/ml)'].hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# add in previusly-repirted data from JGI paper \n",
    "\n",
    "import pandas \n",
    "\n",
    "fr = pandas.read_csv('../data_sets/freezer_annotated.csv', index_col=0) \n",
    "\n",
    "color_map = {\n",
    "    'Bacteria': '#51A7F9', # blue \n",
    "    'Eukaryote': '#70BF41', # green \n",
    "    'Archaea': '#EC5D57'} # red \n",
    "\n",
    "def shape_map(value):\n",
    "    if value < 0.1:\n",
    "        # did not express\n",
    "        return 'd'\n",
    "    return 'o'\n",
    "\n",
    "collect = []\n",
    "for idx, row in fr.iterrows():\n",
    "    collect.append('{}\\t{}\\t{}\\n'.format(idx, 'clade_marker_size', '20'))\n",
    "    collect.append('{}\\t{}\\t{}\\n'.format(idx, 'clade_marker_color', color_map[row.Kindgom]))\n",
    "    collect.append('{}\\t{}\\t{}\\n'.format(idx, 'clade_marker_shape', shape_map(row['Sol_exp_(mg/ml)']))) \n",
    "\n",
    "with open('annot.txt', 'w') as fn:\n",
    "    fn.write(''.join(collect))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     525 annot.txt\r\n"
     ]
    }
   ],
   "source": [
    "! wc -l annot.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SDJGI_57\tclade_marker_shape\td\r\n",
      "SDJGI_184\tclade_marker_size\t20\r\n",
      "SDJGI_184\tclade_marker_color\t#51A7F9\r\n",
      "SDJGI_184\tclade_marker_shape\to\r\n",
      "SDJGI_72\tclade_marker_size\t20\r\n",
      "SDJGI_72\tclade_marker_color\t#EC5D57\r\n",
      "SDJGI_72\tclade_marker_shape\to\r\n",
      "SDJGI_116\tclade_marker_size\t20\r\n",
      "SDJGI_116\tclade_marker_color\t#51A7F9\r\n",
      "SDJGI_116\tclade_marker_shape\to\r\n"
     ]
    }
   ],
   "source": [
    "! tail annot.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     551 annot.txt\r\n"
     ]
    }
   ],
   "source": [
    "custom_annot = '''\n",
    "clade_marker_size\t0\n",
    "clade_separation\t0.3\n",
    "\n",
    "SDJGI_142\tclade_marker_size\t100\n",
    "SDJGI_142\tclade_marker_color\tyellow\n",
    "SDJGI_139\tclade_marker_size\t100\n",
    "SDJGI_139\tclade_marker_color\tyellow\n",
    "SDJGI_4\tclade_marker_size\t100\n",
    "SDJGI_4\tclade_marker_color\tyellow\n",
    "SDJGI_16_P22505_BGLB_PAEPO\tclade_marker_size\t500\n",
    "SDJGI_16_P22505_BGLB_PAEPO\tclade_marker_color\tyellow\n",
    "\n",
    "'''\n",
    "\n",
    "with open('annot.txt', 'a') as fn:\n",
    "    fn.write(custom_annot)\n",
    "    \n",
    "! wc -l annot.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "# create XML file \n",
    "\n",
    "export python=/Users/alex/anaconda3/envs/snakes/bin/python \n",
    "$python ../trees/graphlan/graphlan_annotate.py --annot annot.txt mafft_tree.nwk mafft_graphlan.xml  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "# create XML file \n",
    "\n",
    "export python=/Users/alex/anaconda3/envs/snakes/bin/python \n",
    "$python graphlan.py --external_legends --pad 0 --dpi 200 mafft_graphlan.xml output_img.png\n",
    "open output_img.png"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
